Improve EmbeddingBag backward perf (esp. when grad being non-contiguous)

bddppq opened the pull request:
Summary:
Benchmark:

https://gist.github.com/bddppq/508555897a185ad9a85574eed00b93f6

```
Run on (24 X 2394.42 MHz CPU s)
2019-12-02 16:02:43
-----------------------------------------------------------------------------
Benchmark                                      Time           CPU Iterations
-----------------------------------------------------------------------------
Contiguous<64>/index_64                     3877 ns       3871 ns     187864
Contiguous<64>/index_select_64               966 ns        966 ns     751247
NonContiguous<64>/index_64                  5657 ns       5657 ns     121703
NonContiguous<64>/index_select_64          30809 ns      30809 ns      22619
Contiguous<128>/index_128                   8305 ns       8304 ns      84927
Contiguous<128>/index_select_128           13305 ns      13172 ns      45040
NonContiguous<128>/index_128               10663 ns      10657 ns      69002
NonContiguous<128>/index_select_128        63543 ns      63543 ns      11409
Contiguous<256>/index_256                  13924 ns      13918 ns      50214
Contiguous<256>/index_select_256           34822 ns      34822 ns      19851
NonContiguous<256>/index_256               46786 ns      46782 ns      15710
NonContiguous<256>/index_select_256       157534 ns     157533 ns       4418
Contiguous<512>/index_512                  37008 ns      36998 ns      19786
Contiguous<512>/index_select_512           74236 ns      74208 ns       9283
NonContiguous<512>/index_512              195606 ns     195488 ns       3614
NonContiguous<512>/index_select_512       412587 ns     412502 ns       1665
Contiguous<1024>/index_1024               105950 ns     105907 ns       6639
Contiguous<1024>/index_select_1024        181358 ns     181296 ns       3802
NonContiguous<1024>/index_1024            750279 ns     750090 ns        965
NonContiguous<1024>/index_select_1024    1229192 ns    1228963 ns        574
Contiguous<2048>/index_2048               392678 ns     392661 ns       1724
Contiguous<2048>/index_select_2048        554955 ns     554725 ns       1279
NonContiguous<2048>/index_2048          10443905 ns   10441740 ns         73
NonContiguous<2048>/index_select_2048   11573477 ns   11570427 ns         59
```

Note https://github.com/pytorch/pytorch/pull/30598 is working on improving the perf of `index_select`. Once that PR is landed we should re-run the benchmark to see if we should switch back.

Test Plan: waitforsandcastle

Differential Revision: D18705593

colesbury commented:
1. I think this code path is used by GPU EmbeddingBag backward as well. Can you benchmark that to make sure there's no regression?

2. I'm a little concerned about the regression for the 64x64 contiguous case. Can you investigate this a bit more? For example, keep the columns of `input` fixed as 64 but increase the number of rows. (64 is a plausible number of columns but seems really small for the number of rows)

bddppq commented:
@colesbury Good point. I re-ran the benchmarks for more different combinations of (#row, #col) and found for contiguous input with small #col, `index_select` is consistently faster (the tipping point is 64), for other cases `index` is faster.

Benchmark code: https://gist.github.com/bddppq/216fc8abfa8b68fa6b861974e977ded6

![image](https://user-images.githubusercontent.com/9300575/70105141-e405a480-15f4-11ea-95e2-8c26fa4188e1.png)
![image](https://user-images.githubusercontent.com/9300575/70105150-e962ef00-15f4-11ea-8fee-26bf155228a2.png)
![image](https://user-images.githubusercontent.com/9300575/70105161-f253c080-15f4-11ea-8d05-227c4b6cc847.png)
![image](https://user-images.githubusercontent.com/9300575/70105180-fe3f8280-15f4-11ea-8d07-b4741a108638.png)
![image](https://user-images.githubusercontent.com/9300575/70105231-131c1600-15f5-11ea-9fa8-5d5825167f77.png)
![image](https://user-images.githubusercontent.com/9300575/70105206-08618100-15f5-11ea-9dcf-6b18e67c3c40.png)
![image](https://user-images.githubusercontent.com/9300575/70105245-1adbba80-15f5-11ea-8ac7-38328ecd1ad8.png)
![image](https://user-images.githubusercontent.com/9300575/70105283-3cd53d00-15f5-11ea-90f7-1ea0dbdffebf.png)
![image](https://user-images.githubusercontent.com/9300575/70105292-41015a80-15f5-11ea-94f1-a6a84e2d97ea.png)
![image](https://user-images.githubusercontent.com/9300575/70105297-452d7800-15f5-11ea-889a-9215a0618bde.png)
![image](https://user-images.githubusercontent.com/9300575/70105302-48286880-15f5-11ea-9ced-e4913c638ee3.png)
![image](https://user-images.githubusercontent.com/9300575/70105315-4bbbef80-15f5-11ea-8c75-c6d0e7940c67.png)
![image](https://user-images.githubusercontent.com/9300575/70105318-4fe80d00-15f5-11ea-9afc-45e2b9c540fa.png)
![image](https://user-images.githubusercontent.com/9300575/70105355-6aba8180-15f5-11ea-9d9d-830df5fe93dd.png)
![image](https://user-images.githubusercontent.com/9300575/70105360-6e4e0880-15f5-11ea-8431-886e64734327.png)
![image](https://user-images.githubusercontent.com/9300575/70105364-7148f900-15f5-11ea-9b40-9810b44cf43b.png)
![image](https://user-images.githubusercontent.com/9300575/70105368-74dc8000-15f5-11ea-9887-4c8e142322ea.png)
![image](https://user-images.githubusercontent.com/9300575/70105376-79a13400-15f5-11ea-87e5-b0e311dc55dd.png)

bddppq committed:
Improve EmbeddingBag backward perf (esp. when grad being non-contiguous) (#30639)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30639

Benchmark:

https://gist.github.com/bddppq/216fc8abfa8b68fa6b861974e977ded6

Note https://github.com/pytorch/pytorch/pull/30598 is working on improving the perf of `index_select`. Once that PR is landed we should re-run the benchmark to see if we should switch back.

Test Plan: waitforsandcastle

Differential Revision: D18705593

fbshipit-source-id: ef9fad99378cf1db599164e89780d128406ec3cc
